handle/Timer.lua

```diff
@@ -1,30 +1,190 @@
+	end
+	Native.PauseTimer(t)
+end)
+
+replaceNative('ResumeTimer', function(t)
+	if not Timer[t].hasCallback then
+		Log.error("ResumeTimer: Attempted to resume a timer " .. I2S(GetHandleId(t)) .. " with no callback")
+		return
+	end
+	if Timer[t].inCallback and Timer[t].looped then
+		return
+	end
+	if not Timer[t].inCallback and not Timer[t].running then
+		Timer[t].running    = 0
+	end
+	Native.ResumeTimer(t)
+end)
+
+replaceNative('TimerStart', function(t, dur, looper, func)
+	if not Timer[t].hasCallback then
+		--  Make it come true
+		Timer[t].hasCallback = 0
+	end
+	if not Timer[t].inCallback then
+		--  Usually when a timer is created
+		Timer[t].duration   = math.max(dur, 0.)
+		Timer[t].looped     = looper
+		Timer[t].callback   = func
+		Timer[t].running    = 0
+
+		--  Create a new function that will act as the callback
+		Native.timerStart(t, dur, looper, function()
+			local tr = GetExpiredTimer()
+   
+			Timer[tr].inCallback = 0
+			Timer[tr].running    = nil
+			Timer[tr].callback()
+   
+			Timer[tr].inCallback = nil
+			if Timer[tr].destroyFlag then
+				DestroyTimer(tr)
+			elseif Timer[tr].tempData then
+				--  Properties of the timer were overwritten
+				Native.PauseTimer(tr)
+				TimerStart(tr, Timer[tr].tempData.dur, Timer[tr].tempData.looper, Timer[tr].tempData.func)
+				Timer[tr].tempData = nil
+				Timer[tr].elapsed = 0.
+			else
+				if Timer[tr].looped and ((BlzBitAnd(Timer[tr].pauseFlag, 1) ~= 0) or (BlzBitAnd(Timer[tr].pauseFlag, 2) ~= 0)) then
+					Timer[tr].pauseFlag  = BlzBitAnd(Timer[tr].pauseFlag, 0)
+					Native.PauseTimer(tr)
+					TimerStart(tr, Timer[tr].duration, Timer[tr].looped, Timer[tr].callback)
+				elseif Timer[tr].looped then
+					Timer[tr].running    = 0
+				end
+				Timer[tr].elapsed = 0.
+			end
+		end)
 	else
-		return Native.CreateTimer()
+		if not Timer[t].tempData then
+			Timer[t].tempData = {__mode='k'}
+		end
+		Timer[t].tempData.dur = dur
+		Timer[t].tempData.looper = looper
+		Timer[t].tempData.func = func
 	end
 end)
 
 replaceNative('DestroyTimer', function(t)
-	if t == nil then
-		Log.error('DestroyTimer: null timer')
+	if not Timer[t].hasCallback then
+		Timer[t]    = nil
+		free_timers.push(t)
 		return
 	end
-	if free_timers[t] then
-		Log.error('DestroyTimer: double free')
+	if not Timer[t] or Timer[t].inCallback then
+		--  No need to destroy an already-destroyed timer
+		if Timer[t].inCallback then
+			Timer[t].onDestroyFlag  = 0
+		end
 		return
 	end
-	PauseTimer(t)
-	free_timers[t] = true
+	if Timer[t].onDestroyFlag then
+		Timer[t].onDestroyFlag = nil
+	end
+	if Timer[t].running or Timer[t].looped then
+		Timer[t].running = nil
+		Native.PauseTimer(t)
+	end
+	Timer[t]    = nil
+	free_timers.push(t)
 end)
 
+replaceNative('TimerGetTimeout', function(t) return Timer[t].duration end)
+replaceNative('TimerGetElapsed', function(t) return Native.TimerGetElapsed(t) + Timer[t].elapsed end)
+
+function TimerSetRemaining(t, newR, update)
```

Comment:
Do we need this function at all? Considering the now complexity of your native replacements, it could be worth itself, but I wonder if you could simplify everything so this won't be needed.